### PR TITLE
AFS: Add unit tests for scheduling logic with AFS

### LINKED ISSUE #6248: AFS: add unit tests for scheduling logic with AFS
**What would you like to be cleaned**:

Add unit tests for scheduling with AFS which includes entry penalties.

It could be a new test function like "TestScheduleForAFS", by analogy to "TestScheduleForTAS", because the main "TestSchedule" is already enormous. 

**Why is this needed**:

They were not added in the main PR: https://github.com/kubernetes-sigs/kueue/pull/6189

Unit tests are very useful for the ease of debugging. Unit tests are also much faster to run than integration tests, which already take 20min, so they allow to cover many corner cases almost cost free.

### PR DESCRIPTION
**Pull request by IrvingMg**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Add unit tests.

#### Which issue(s) this PR fixes:

Fixes #6248 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -9345,3 +9348,297 @@ func TestScheduleForTASCohorts(t *testing.T) {
...
+		afsConfig.UsageHalfLifeTime.Seconds(),
+	)
+
+	result := corev1.ResourceList{}
+	for _, request := range requests {
+		penalty := utilresource.MulByFloat(request, alpha)
+		result = utilresource.MergeResourceListKeepSum(result, penalty)
+	}
+	return result
+}
```
Comment (PBundyra) [inline pkg/scheduler/scheduler_test.go]: Let's not use the logic we want to test in the tests themselves

File: pkg/scheduler/scheduler_test.go [hunk 2]
```diff
@@ -9345,3 +9348,297 @@ func TestScheduleForTASCohorts(t *testing.T) {
...
+						Request(corev1.ResourceCPU, "4").
+						Obj()).
+					Obj(),
+			},
+			wantScheduled: []workload.Reference{"default/wl1"},
+			wantAssignments: map[workload.Reference]kueue.Admission{
+				"default/wl1": *utiltesting.MakeAdmission("cq1", "one").Assignment(corev1.ResourceCPU, "default", "4").Obj(),
+			},
+			wantPenalties: map[string]corev1.ResourceList{
+				"default/lq-a": calculateExpectedPenalty([]corev1.ResourceList{
```
Comment (PBundyra) [inline pkg/scheduler/scheduler_test.go]: Please fill with a const number for every scenario

File: pkg/scheduler/scheduler_test.go [hunk 3]
```diff
@@ -9345,3 +9348,297 @@ func TestScheduleForTASCohorts(t *testing.T) {
...
+
+			ctx, log := utiltesting.ContextWithLog(t)
+			for _, q := range queues {
+				if err := qManager.AddLocalQueue(ctx, &q); err != nil {
+					t.Fatalf("Inserting queue %s/%s in manager: %v", q.Namespace, q.Name, err)
+				}
+			}
+			for i := range resourceFlavors {
+				cqCache.AddOrUpdateResourceFlavor(log, resourceFlavors[i])
+			}
```
Comment (PBundyra) [inline pkg/scheduler/scheduler_test.go]:
```suggestion
			for _, rf := range resourceFlavors {
				cqCache.AddOrUpdateResourceFlavor(log, rf)
			}
```

Just for consistency

File: pkg/scheduler/scheduler_test.go [hunk 4]
```diff
@@ -9345,3 +9348,297 @@ func TestScheduleForTASCohorts(t *testing.T) {
...
+			ClusterQueue("cq1").Obj(),
+		*utiltesting.MakeLocalQueue("lq-b", "default").
+			FairSharing(&kueue.FairSharing{Weight: ptr.To(resource.MustParse("1"))}).
+			ClusterQueue("cq1").Obj(),
+	}
+
+	cases := map[string]struct {
+		workloads       []kueue.Workload
+		admissionError  error
+		wantScheduled   []workload.Reference
```
Comment (PBundyra) [inline pkg/scheduler/scheduler_test.go]: This seems redundant as we have this information in `wantAssignments` field

File: pkg/scheduler/scheduler_test.go (hunk 3 above)
Comment (IrvingMg) [reply pkg/scheduler/scheduler_test.go]: Done.

File: pkg/scheduler/scheduler_test.go (hunk 2 above)
Comment (IrvingMg) [reply pkg/scheduler/scheduler_test.go]: Done.

File: pkg/scheduler/scheduler_test.go (hunk 1 above)
Comment (IrvingMg) [reply pkg/scheduler/scheduler_test.go]: Got it. Replaced with constant values.

File: pkg/scheduler/scheduler_test.go (hunk 4 above)
Comment (IrvingMg) [reply pkg/scheduler/scheduler_test.go]: Right. I've removed it.

File: pkg/scheduler/scheduler_test.go [hunk 5]
```diff
@@ -9345,3 +9347,282 @@ func TestScheduleForTASCohorts(t *testing.T) {
...
+						continue
+					}
+
+					expectedFloat := utilresource.QuantityToFloat(&expectedQuantity)
+					actualFloat := utilresource.QuantityToFloat(&actualQuantity)
+					if diff := cmp.Diff(expectedFloat, actualFloat, cmpopts.EquateApprox(0.01, 0)); diff != "" {
+						t.Errorf("Unexpected penalty for %s resource %s (-want,+got):\n%s", lqKey, resource, diff)
+					}
+				}
+			}
```
Comment (PBundyra) [inline pkg/scheduler/scheduler_test.go]: Why don't we just compare `gotPenalties` with `tc.wantPenalties`? The results should always be the same as they're not based on reconcile time but constant values (halflife time and sampling)

Comment (IrvingMg) [reply pkg/scheduler/scheduler_test.go]: The problem is that we have resource.Quantity type for penalties and also using resource.MulByFloat method for calculation which is causing rounding differences. Then we need to handle those cases giving some margin for precision.

Comment (PBundyra) [reply pkg/scheduler/scheduler_test.go]: Where does the difference in rounding come from if the input is always the same?

Comment (IrvingMg) [reply pkg/scheduler/scheduler_test.go]: I think the difference comes from this part of the code: https://github.com/kubernetes-sigs/kueue/blob/36928e70fc50238b5b10ea2ed9c6391bfd399542/pkg/util/resource/resource.go#L110-L117

For the alpha calculation, the input is the same. However, for the penalty calculation, we also use the request values, which are different, and we’re using `MulByFloat`. Line 114 truncates the values, which causes small differences that require approximate comparison in tests.

Comment (mimowo) [reply pkg/scheduler/scheduler_test.go]: What are the benefits of asserting on penalties, it seems we could consider them as a technical detail from the schedule() perspective.


WDYT?

Comment (PBundyra) [reply pkg/scheduler/scheduler_test.go]: > I think the difference comes from this part of the code:
> 
> https://github.com/kubernetes-sigs/kueue/blob/36928e70fc50238b5b10ea2ed9c6391bfd399542/pkg/util/resource/resource.go#L110-L117
> 
> For the alpha calculation, the input is the same. However, for the penalty calculation, we also use the request values, which are different, and we’re using `MulByFloat`. Line 114 truncates the values, which causes small differences that require approximate comparison in tests.

Okay but the output of the rounding is always the same for a given scenario. If e.g. the exact result of the calculation is `0.335`, and our code outputs  `0.334`, then we should assert against `0.334` with the comment it was rounded `0.335 ~ 0.334`

Comment (PBundyra) [reply pkg/scheduler/scheduler_test.go]: > What are the benefits of asserting on penalties, it seems we could consider them as a technical detail from the schedule() perspective.
> 
> WDYT?

Do you mean we should only test which workloads are admitted/not admitted?

Comment (mimowo) [reply pkg/scheduler/scheduler_test.go]: Well, that is a suggestion to consider. 

I think it is reasonable to start simple, and only add asserting on penalties if we have scenarios which cannot be tested otherwise in unit tests. Such very strong asserts on internal calculations may sometimes make refactoring problematic.

For example I love this test: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/scheduler/scheduler_test.go#L1631-L1661

It asserts workload scheduling, but the comment describes the test intention, and outlines the internal calculations to understand the test scenario.

Comment (PBundyra) [reply pkg/scheduler/scheduler_test.go]: I like that, let's start with scenarios similar to the ones in integration tests

Comment (mimowo) [reply pkg/scheduler/scheduler_test.go]: Yes, you may consider more test cases, as the minimum I would like 3 with the following properties:
- passes regardless if AFS enabled just to show AFS is not breaking anything
- fails without AFS, passes regardless of penalties
- only passing with AFS with penalties

Comment (IrvingMg) [reply pkg/scheduler/scheduler_test.go]: I've updated the test cases. PTAL.

**Comment by mimowo** [comment]
/approve
leaving lgtm to @PBundyra

**Comment by PBundyra** [comment]
/lgtm
Thanks @IrvingMg
